rkt/fly.go

```diff
@@ -0,0 +1,402 @@
+	if err != nil {
+		stderr("fly: error preparing App: %v", err)
+		return 1
+	}
+
+	app := rktApps.Last()
+	var execargs []string
+	if flagExec != "" {
+		log.Printf("Overriding exec with %q", flagExec)
+		execargs = []string{flagExec}
+	} else {
+		execargs = imApp.Exec
+	}
+	execargs = append(execargs, app.Args...)
+
+	rfs := filepath.Join(dir, "rootfs")
+	if _, err := os.Stat(rfs); os.IsNotExist(err) {
+		stderr("fly: target root directory %q", rfs, err)
+		os.RemoveAll(dir)
+		return 1
+	}
+
+	type vmTuple struct {
+		V types.Volume
+		M schema.Mount
+	}
+
+	namedVolumeMounts := map[types.ACName]vmTuple{}
+
+	// Insert the command-line Mounts
+	for _, m := range rktApps.Mounts {
+		_, exists := namedVolumeMounts[m.Volume]
+		if exists {
+			stderr("fly: duplicated mount given: %q", m.Volume)
+			os.RemoveAll(dir)
+			return 1
+		}
+		namedVolumeMounts[m.Volume] = vmTuple{M: m}
+	}
+
+	// Merge command-line Mounts with ImageManifest's MountPoints
+	for _, mp := range imApp.MountPoints {
+		tuple, exists := namedVolumeMounts[mp.Name]
+		switch {
+		case exists && tuple.M.Path != mp.Path:
+			stderr("fly: conflicting path information from mount and mountpoint %q", mp.Name)
+			os.RemoveAll(dir)
+			return 1
+		case !exists:
+			namedVolumeMounts[mp.Name] = vmTuple{M: schema.Mount{Volume: mp.Name, Path: mp.Path}}
+		}
+	}
+
+	// Insert the command-line Volumes
+	for _, v := range rktApps.Volumes {
+		// Check if we have a mount for this volume
+		tuple, exists := namedVolumeMounts[v.Name]
+		if !exists {
+			stderr("fly: missing mount for volume %q", v.Name)
+			os.RemoveAll(dir)
+			return 1
+		} else if tuple.M.Volume != v.Name {
+			// TODO(steveeJ): remove this case. it's merely a safety mechanism regarding the implementation
+			stderr("fly: mismatched volume:mount pair: %q != %q", v.Name, tuple.M.Volume)
+			os.RemoveAll(dir)
+			return 1
+		}
+		namedVolumeMounts[v.Name] = vmTuple{V: v, M: tuple.M}
+	}
+
+	// Merge command-line Volumes with ImageManifest's MountPoints
+	for _, mp := range imApp.MountPoints {
+		// Check if we have a volume for this mountpoint
+		tuple, exists := namedVolumeMounts[mp.Name]
+		if !exists || tuple.V.Name == "" {
+			stderr("fly: missing volume for mountpoint %q", mp.Name)
+			os.RemoveAll(dir)
+			return 1
+		}
+
+		// If empty, fill in ReadOnly bit
+		if tuple.V.ReadOnly == nil {
+			v := tuple.V
+			v.ReadOnly = &mp.ReadOnly
+			namedVolumeMounts[mp.Name] = vmTuple{M: tuple.M, V: v}
+		}
+	}
+
+	argFlyMounts := []flyMount{}
+	for _, tuple := range namedVolumeMounts {
+		var flags uintptr = syscall.MS_BIND | syscall.MS_REC
+		if tuple.V.ReadOnly != nil && *tuple.V.ReadOnly {
+			flags |= syscall.MS_RDONLY
+		}
+		argFlyMounts = append(
+			argFlyMounts,
+			flyMount{tuple.V.Source, rfs, tuple.M.Path, "none", flags},
+		)
+	}
+
+	// create a separate mount namespace so the filesystems
+	// are unmounted when exiting the pod
+	if err := syscall.Unshare(syscall.CLONE_NEWNS); err != nil {
```

Comment:
Unshare works on the thread level here, so the following mount can still happen in host mount namespace.
But here we can use LockOSThread to make the mounting happen in the unshared thread. But since we don't have controls over the creation of goroutines, it can cause new threads to be cloned from this unshared threads. though it doesn't seem to cause a problem here.

But anyway it is unrecommended to use unshare in go :(  https://github.com/golang/go/issues/1954